plugins/infinite-sky.cpp

```diff
@@ -0,0 +1,270 @@
+
+#include "Core.h"
+#include "DataDefs.h"
+#include "Debug.h"
+#include "Export.h"
+#include "LuaTools.h"
+#include "PluginManager.h"
+
+#include "modules/World.h"
+#include "modules/EventManager.h"
+
+#include "df/construction.h"
+#include "df/map_block.h"
+#include "df/map_block_column.h"
+#include "df/world.h"
+#include "df/z_level_flags.h"
+
+#include <cstring>
+#include <string>
+#include <vector>
+
+using namespace std;
+
+using namespace DFHack;
+using namespace df::enums;
+
+DFHACK_PLUGIN("infinite-sky");
+DFHACK_PLUGIN_IS_ENABLED(is_enabled);
+
+REQUIRE_GLOBAL(world);
+
+namespace DFHack {
+    // for configuration-related logging
+    DBG_DECLARE(infiniteSky, control, DebugCategory::LINFO);
+    // for logging during creation of z-levels
+    DBG_DECLARE(infiniteSky, cycle, DebugCategory::LINFO);
+}
+
+static const string CONFIG_KEY = string(plugin_name) + "/config";
+static PersistentDataItem config;
+enum ConfigValues {
+    CONFIG_IS_ENABLED = 0,
+};
+
+command_result infiniteSky (color_ostream &out, std::vector <std::string> & parameters);
+
+static void constructionEventHandler(color_ostream& out, void* ptr);
+EventManager::EventHandler handler(plugin_self, constructionEventHandler,0);
+
+DFhackCExport command_result plugin_init(color_ostream &out,
+                                         std::vector<PluginCommand> &commands) {
+    commands.push_back(PluginCommand(
+        "infinite-sky", "Automatically allocate new z-levels of sky.",
+        infiniteSky));
+    return CR_OK;
+}
+
+void cleanup() {
+    EventManager::unregister(EventManager::EventType::CONSTRUCTION, handler);
+}
+
+DFhackCExport command_result plugin_enable(color_ostream &out, bool enable) {
+    if (!Core::getInstance().isMapLoaded() || !World::isFortressMode()) {
+        out.printerr("Cannot enable %s without a loaded fort.\n", plugin_name);
+        return CR_FAILURE;
+    }
+    if (enable != is_enabled) {
+        is_enabled = enable;
+        DEBUG(control, out)
+            .print("%s from the API; persisting\n",
+                   is_enabled ? "enabled" : "disabled");
+        config.set_bool(CONFIG_IS_ENABLED, is_enabled);
+
+        if (enable) {
+            EventManager::registerListener(
+                EventManager::EventType::CONSTRUCTION, handler);
+        } else {
+            cleanup();
+        }
+    } else {
+        DEBUG(control, out)
+            .print("%s from the API, but already %s; no action\n",
+                   is_enabled ? "enabled" : "disabled",
+                   is_enabled ? "enabled" : "disabled");
+    }
+
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_shutdown(color_ostream &out) {
+    cleanup();
+    return CR_OK;
+}
```

Comment:
listener cleanup on shutdown is managed by the framework and you don't need to handle it here.